lib/logstash/filters/hashid.rb

```diff
@@ -0,0 +1,121 @@
+
+  # Target field.
+  # Will overwrite current value of a field if it exists.
+  config :target, :validate => :string, :default => 'hashid'
+
+  # Encryption key to be used when generating cryptographic hashes
+  config :key, :validate => :string, :default => 'hashid'
+
+  # Hash function to use
+  config :method, :validate => ['SHA1', 'SHA256', 'SHA384', 'SHA512', 'MD5'], :default => 'MD5'
+
+  # If full hash generated is not to be used, this parameter specifies how many bytes that should be used
+  # If not specified, the full hash will be used
+  config :hash_bytes_used, :validate => :number
+
+  # Use the timestamp to generate an ID prefix
+  config :add_timestamp_prefix, :validate => :boolean, :default => true
+
+  CHARS = "-0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz".freeze
+
+  def register
+    # convert to symbol for faster comparisons
+    @method = @method.to_sym
+    @digest = select_digest(@method)
+  end
+
+  def filter(event)
+    hmac = OpenSSL::HMAC.new(@key, @digest.new)
+
+    @source.sort.each do |k|
+      hmac.update("|#{k}|#{event[k]}") 
+    end
+
+    hash = hmac.digest
+
+    if !@hash_bytes_used.nil? && @hash_bytes_used > 0 && hash.length > @hash_bytes_used
+      hash = hash[(-1 * @hash_bytes_used), @hash_bytes_used]
+    end
+
+    epoch_array = []
+    if @add_timestamp_prefix
+      epoch = event[@timestamp_field].to_i
+      epoch_array.push(epoch >> 24)
+      epoch_array.push((epoch >> 16) % 256)
+      epoch_array.push((epoch >> 8) % 256)
+      epoch_array.push(epoch % 256)
+    end
+
+    binary_array = epoch_array + hash.unpack('C*')
+
+    event[@target] = encode_to_sortable_string(binary_array).force_encoding(Encoding::UTF_8)
+  end
+
+  def select_digest(method)
+    case method
+    when :SHA1
+      OpenSSL::Digest::SHA1.new
+    when :SHA256
+      OpenSSL::Digest::SHA256.new
+    when :SHA384
+      OpenSSL::Digest::SHA384.new
+    when :SHA512
+      OpenSSL::Digest::SHA512.new
+    when :MD5
+      OpenSSL::Digest::MD5.new
+    else
+      # we really should never get here
+      raise(LogStash::ConfigurationError, "Unknown digest for method=#{method.to_s}")
+    end
+  end
+
+  def encode_to_sortable_string(data)
+    encoded_string = ""
+    offset = 0
+    while offset < data.length do
+      buf = data[offset,3]
+      offset+=3
+
+      pad = '' # padding for this group of 4 characters
+      while buf.length < 3
+        buf.push(0)
+        pad << '='
+      end
+
+      group24 = (buf[0] << 16) | (buf[1] << 8) | buf[2] # current 3 bytes as a 24 bit value
+      encoded = CHARS[(group24 >> 18) & 0x3f, 1] # read the 24 bit value 6 bits at a time
+      encoded << CHARS[(group24 >> 12) & 0x3f, 1]
+      encoded << CHARS[(group24 >> 6) & 0x3f, 1]
+      encoded << CHARS[(group24 >> 0) & 0x3f, 1]
+      encoded[4 - pad.length, pad.length] = pad # add the padding
+      encoded_string << encoded
+    end
+
+    encoded_string.tr('=','')
+  end
```

Comment:
Consider this alternative implementation - it uses numbers and arrays until the end when it builds the final string once.

```ruby
  CHARS = '-0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz='.chars.to_a.freeze
  SHIFTS = [18, 12, 6, 0].freeze


  def encode_to_sortable_string(data)
    idxes = []
    to_take = 0
    data.each_slice(3) do |part0, part1, part2|
      to_take = 0
      if part1.nil?
        part1 = part2 = 0
        to_take = 2
      end
      if part2.nil?
        part2 = 0
        to_take = 1
      end
      group24 = (part0 << 16) | (part1 << 8) | part2
      idxes.concat(SHIFTS.map{|n| (group24 >> n) & 0x3f })
    end
    CHARS.values_at(*idxes.take(idxes.size - to_take)).join
  end
```
 I have benchmarked this (as a2) against the original (as a2).
```
Benchmark.ips do |x|
  bmi = EncodeBench.new
  x.config(:time => 5, :warmup => 20)

  x.report("a2")      { bmi.encode_to_sortable_string2 }
  x.report("a1")      { bmi.encode_to_sortable_string1 }

  x.compare!
end
Warming up --------------------------------------
                  a2    14.387k i/100ms
                  a1    11.663k i/100ms
Calculating -------------------------------------
                  a2    159.028k (± 2.3%) i/s -    805.672k
                  a1    125.590k (± 2.5%) i/s -    629.802k

Comparison:
                  a2:   159028.1 i/s
                  a1:   125590.0 i/s - 1.27x slower

=> #<Benchmark::IPS::Report:0x15dd5ac2 @entries=[#<Benchmark::IPS::Report::Entry:0x7bbbb6a8 @label="a2", @ips=159028.10691169664, @show_total_time=false, @iterations=805672, @measurement_cycle=14387, @microseconds=5069000.005722046, @ips_sd=3634>, #<Benchmark::IPS::Report::Entry:0x58a120b0 @label="a1", @ips=125589.97272417996, @show_total_time=false, @iterations=629802, @measurement_cycle=11663, @microseconds=5018000.12588501, @ips_sd=3104>], @data=nil>

```